spark3/src/test/java/org/apache/iceberg/spark/source/TestStructuredStreamingRead3.java

```diff
@@ -0,0 +1,502 @@
+
+    // this should create a snapshot with type delete.
+    table.newDelete()
+        .deleteFromRowFilter(Expressions.equal("id", 4))
+        .commit();
+
+    // check pre-condition - that the above newDelete operation on table resulted in Snapshot of Type DELETE.
+    Assert.assertEquals(DataOperations.DELETE, table.currentSnapshot().operation());
+
+    try {
+      Dataset<Row> df = spark.readStream()
+          .format("iceberg")
+          .load(location.toString());
+      StreamingQuery streamingQuery = df.writeStream()
+          .format("memory")
+          .queryName("testtablewithdelete")
+          .outputMode(OutputMode.Append())
+          .start();
+
+      try {
+        streamingQuery.processAllAvailable();
+        Assert.assertTrue(false); // should be unreachable
+      } catch (Exception exception) {
+        Assert.assertTrue(exception instanceof StreamingQueryException);
+        Assert.assertTrue(((StreamingQueryException) exception).cause() instanceof IllegalStateException);
+      }
+
+    } finally {
+      for (StreamingQuery query : spark.streams().active()) {
+        query.stop();
+      }
+    }
+  }
+
+  private static List<SimpleRecord> processMicroBatch(DataStreamWriter<Row> singleBatchWriter, String viewName)
+      throws TimeoutException, StreamingQueryException {
+    StreamingQuery streamingQuery = singleBatchWriter.start();
+    streamingQuery.awaitTermination();
+
+    return spark.sql(String.format("select * from %s", viewName))
+        .as(Encoders.bean(SimpleRecord.class))
+        .collectAsList();
+  }
+
+  /**
+   * get test data - a list of records per snapshot
+   */
+  private static List<List<SimpleRecord>> getTestDataForMultipleSnapshots() {
+    return Lists.newArrayList(
+        Lists.newArrayList(
+            new SimpleRecord(1, "one"),
+            new SimpleRecord(2, "two"),
+            new SimpleRecord(3, "three")),
+        Lists.newArrayList(
+            new SimpleRecord(4, "four"),
+            new SimpleRecord(5, "five")),
+        Lists.newArrayList(
+            new SimpleRecord(6, "six"),
+            new SimpleRecord(7, "seven"))
+    );
+  }
+
+  /**
+   * appends each list as a Snapshot on the iceberg table at the given location.
+   * accepts a list of lists - each list representing data per snapshot.
+   */
+  private static void appendData(List<List<SimpleRecord>> data, File location) {
+    // generate multiple snapshots
+    for (List<SimpleRecord> l : data) {
+      appendData(l, location, "parquet");
+    }
+  }
+
+  private static void appendData(List<SimpleRecord> data, File location, String fileFormat) {
+    Dataset<Row> df = spark.createDataFrame(data, SimpleRecord.class);
+    df.select("id", "data").write()
+        .format("iceberg")
+        .option("write-format", fileFormat)
+        .mode("append")
+        .save(location.toString());
+  }
+
+  private static List<SimpleRecord> processStreamTillEnd(Dataset<Row> df) throws TimeoutException {
```

Comment:
Looks like this method name is slightly misleading because it isn't processing and ending the stream, it is processing all of the available data. How about `processAvailable`?